src/passes/Poppify.cpp

```diff
@@ -0,0 +1,504 @@
+}
+
+void Poppifier::emitLocalSet(LocalSet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->value->type.isTuple()) {
+    auto types = func->getLocalType(curr->index);
+    const auto& elems = tupleVars[curr->index];
+    // Add the unconditional sets
+    for (size_t i = types.size() - 1; i >= 1; --i) {
+      instrs.push_back(
+        builder.makeLocalSet(elems[i], builder.makePop(types[i])));
+    }
+    if (curr->isTee()) {
+      // Use a tee followed by gets to retrieve the tuple
+      instrs.push_back(
+        builder.makeLocalTee(elems[0], builder.makePop(types[0]), types[0]));
+      for (size_t i = 1; i < types.size(); ++i) {
+        instrs.push_back(builder.makeLocalGet(elems[i], types[i]));
+      }
+    } else {
+      // Otherwise just add the last set
+      instrs.push_back(
+        builder.makeLocalSet(elems[0], builder.makePop(types[0])));
+    }
+  } else {
+    poppify(curr);
+    instrs.push_back(curr);
+  }
+}
+
+void Poppifier::emitGlobalGet(GlobalGet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->type.isTuple()) {
+    auto types = module->getGlobal(curr->name)->type;
+    for (Index i = 0; i < types.size(); ++i) {
+      instrs.push_back(
+        builder.makeGlobalGet(getGlobalElem(curr->name, i), types[i]));
+    }
+  } else {
+    instrs.push_back(curr);
+  }
+}
+
+void Poppifier::emitGlobalSet(GlobalSet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->value->type.isTuple()) {
+    auto types = module->getGlobal(curr->name)->type;
+    for (Index i = types.size(); i > 0; --i) {
+      instrs.push_back(builder.makeGlobalSet(getGlobalElem(curr->name, i - 1),
+                                             builder.makePop(types[i - 1])));
+    }
+  } else {
+    poppify(curr);
+    instrs.push_back(curr);
+  }
+}
+
+void Poppifier::poppify(Expression* expr) {
+  struct Poppifier : PostWalker<Poppifier> {
+    bool scanned = false;
+    Builder builder;
+    Poppifier(Builder& builder) : builder(builder) {}
+
+    static void scan(Poppifier* self, Expression** currp) {
+      if (!self->scanned) {
+        self->scanned = true;
+        PostWalker<Poppifier>::scan(self, currp);
+      } else {
+        *currp = self->builder.makePop((*currp)->type);
+      }
+    }
+  } poppifier(builder);
+  poppifier.walk(expr);
+}
+
+class PoppifyFunctionsPass : public Pass {
+  bool isFunctionParallel() override { return true; }
+  void
+  runOnFunction(PassRunner* runner, Module* module, Function* func) override {
+    if (func->profile != IRProfile::Poppy) {
+      Poppifier(func, module).write();
+      func->profile = IRProfile::Poppy;
+    }
+  }
+  Pass* create() override { return new PoppifyFunctionsPass; }
+};
+
+} // anonymous namespace
+
+class PoppifyPass : public Pass {
+  void run(PassRunner* runner, Module* module) {
+    PassRunner subRunner(runner);
+    subRunner.add(std::make_unique<PoppifyFunctionsPass>());
+    subRunner.add(std::make_unique<ReFinalize>());
```

Comment:
Why do we need refinalization?

Reply:
Consider this Binaryen IR block:

```
(block $l i32
  (br $l
    (unreachable)
  (i32.const 0)
)
```

In Binaryen IR, this block would have to have type `i32` because it ends in an `i32.const`. However, the `BinaryenIRWriter` would not emit anything after the unreachable, so in Poppy IR we would end up with:

```
(block $l i32
  (unreachable)
)
```

We want to refinalize this block to give it type `unreachable` because there is no reason to keep it typed as `i32`. It's better to give additional optimizations the flexibility and power to do things to this block that do not preserve its `i32` type.